### PR TITLE
RFC for transitional into

### PR DESCRIPTION
**Pull request by anirudhb**
[Rendered](https://github.com/anirudhb/rfcs/blob/master/text/0000-transitional-into.md)

This is my first RFC.

### CODE REVIEW

**Comment by anirudhb** [comment]
Today, this does not work in normal Rust because given this code:
```rust
impl<A, B, C> Into<C> for A
    where A: Into<B>, B: Into<C> {
    fn into(self) -> C {
        self.into().into() 
    }
}
```
It fails with "B is unconstrained"

**Comment by mark-i-m** [comment]
Is the proposal that this should be a blanket impl? Something like

```rust
impl <A, B, C> From<A> for C
where B: From<A>, C: From<A>
{
  fn from(a :A) -> C { from(from(a)) }
}
```

Does anyone know if there are weird coherence issues here?

**Comment by sfackler** [comment]
@mark-i-m there can be multiple choices of B.

**Comment by kennytm** [comment]
Also note that `From` implementation could be lossy and therefore different choices of `B` could lead to different results.

```rust
use std::collections::BinaryHeap;
fn main() {
    let a = vec![4, 12, 8];
    let b = BinaryHeap::from(a);
    let c = Vec::from(b);
    assert_eq!(c, vec![12, 4, 8]);
}
```

File: text/0000-transitional-into.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+[motivation]: #motivation
+
+If we are, e. g. converting units, and we can convert centimeter to meter,
+and meter converts to feet, a conversion from centimeter to feet should be derived.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+If a type `A` implements `Into<B>` and `B` implements `Into<C>`,
+`Into<C>` is automatically implemented for type `A`.
```
[the remark this answers is not in the document]
Comment (anirudhb) [reply text/0000-transitional-into.md]: Good catch. It should drop the derived impl, then.

**Comment by SimonSapin** [comment]
I think that this RFC is just not possible to implement, unfortunately.

As you’ve already found out, the compiler rejects such impls:

```rust
error[E0207]: the type parameter `B` is not constrained by the impl trait, self type, or predicates
   --> libcore/convert.rs:455:9
    |
455 | impl<A, B, C> From<A> for C
    |         ^ unconstrained type parameter

error: aborting due to previous error
```

But even if we were to change the language rules to accept an unconstrained type parameter and force the compiler to do an exhaustive search when trying to determine if this impl is applicable to a given call (which would have compiler performance implications), then we’d only hit the impl coherence rule that would make the trait very impractical at best: this new impl would likely conflict with existing impls. It might even conflict with itself, is multiple choices of `B` would satisfy the bounds.

File: text/0000-transitional-into.md [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+IF another type `D` implements `Into<C>` and `A` implements `Into<D>`,
+the derived implementation will be dropped, because there are two conflicting
```
Comment (sfackler) [inline text/0000-transitional-into.md]: This means that it's a breaking change to add a new `Into` implementation.

**Comment by anirudhb** [comment]
I do agree with the fact that this is nearly impossible to implement, however I'm just throwing the idea out there.

**Comment by sfackler** [comment]
The RFC process is more oriented to things that are possible to implement. A thread on internals.rust-lang.org might be better for something more speculative like this.

**Comment by anirudhb** [comment]
Taking into consideration the issues brought up here, it seems the RFC is impossible to implement. I will be closing this PR.

**Comment by mark-i-m** [comment]
I think many people are experiencing a similar problem, so an internals thread might be a good starting point for discussion :)

**Pull request closed without merging**